x-pack/libbeat/autodiscover/providers/nomad/nomad.go

```diff
@@ -0,0 +1,247 @@
+	}
+
+	builders, err := autodiscover.NewBuilders(config.Builders, config.Hints)
+	if err != nil {
+		return nil, err
+	}
+
+	appenders, err := autodiscover.NewAppenders(config.Appenders)
+	if err != nil {
+		return nil, err
+	}
+
+	metagen, err := nomad.NewMetaGenerator(c, client)
+	if err != nil {
+		return nil, err
+	}
+
+	options := nomad.WatchOptions{
+		SyncTimeout:     config.WaitTime,
+		RefreshInterval: config.SyncPeriod,
+		Node:            config.Host,
+	}
+
+	watcher, err := nomad.NewWatcher(client, options)
+	if err != nil {
+		logp.Err("ERROR creating Watcher %v", err.Error())
+	}
+
+	p := &Provider{
+		config:    config,
+		bus:       bus,
+		uuid:      uuid,
+		templates: mapper,
+		metagen:   metagen,
+		builders:  builders,
+		appenders: appenders,
+		watcher:   watcher,
+	}
+
+	watcher.AddEventHandler(nomad.ResourceEventHandlerFuncs{
+		AddFunc: func(obj nomad.Resource) {
+			logp.Debug("nomad", "Watcher Allocation add: %+v", obj.ID)
+			p.emit(&obj, "start")
+		},
+		UpdateFunc: func(obj nomad.Resource) {
+			logp.Debug("nomad", "Watcher Allocation update: %+v", obj.ID)
+			time.AfterFunc(config.CleanupTimeout, func() { p.emit(&obj, "stop") })
+			time.AfterFunc(config.CleanupTimeout, func() { p.emit(&obj, "start") })
+		},
+		DeleteFunc: func(obj nomad.Resource) {
+			logp.Debug("nomad", "Watcher Allocation delete: %+v", obj.ID)
+			time.AfterFunc(config.CleanupTimeout, func() { p.emit(&obj, "stop") })
+		},
+	})
+
+	return p, nil
+}
+
+// Start for Runner interface.
+func (p *Provider) Start() {
+	if err := p.watcher.Start(); err != nil {
+		logp.Err("Error starting nomad autodiscover provider: %s", err)
+	}
+}
+
+// Stop signals the stop channel to force the watch loop routine to stop.
+func (p *Provider) Stop() {
+	p.watcher.Stop()
+}
+
+// String returns a description of nomad autodiscover provider.
+func (p *Provider) String() string {
+	return "nomad"
+}
+
+func (p *Provider) emit(obj *nomad.Resource, flag string) {
+	// emit one event per allocation with the embedded tasks' metadata
+	nodeName := obj.NodeName
+
+	if len(nodeName) == 0 {
+		// On older versions of Nomad the NodeName property is not set, as a workaround we can use
+		// the NodeID
+		host, err := p.metagen.AllocationNodeName(obj.NodeID)
+		if err != nil {
+			logp.Err("Error fetching node information: %s", err)
+		}
+
+		// If we cannot get a host, we assume that the allocation was stopped
+		if len(host) == 0 {
+			return
+		}
+
+		nodeName = host
+	}
+
+	// common metadata from the entire allocation
+	allocMeta := p.metagen.ResourceMetadata(*obj)
+
+	// job metadatata merged with the task metadata
+	tasks := p.metagen.GroupMeta(obj.Job)
+
+	// emit per-task separated events
+	for _, task := range tasks {
+		event := bus.Event{
+			"provider": p.uuid,
+			"id":       fmt.Sprintf("%s-%s", obj.ID, task["name"]),
+			flag:       true,
+			"host":     nodeName,
+			"meta": common.MapStrUnion(allocMeta, common.MapStr{
+				"task": task,
+			}),
```

Comment:
Good point! 
Perhaps we could avoid adding 2 keys by renaming `meta` to `nomad`? In that regard, it would be similar to the kubernetes provider, something like: `${data.meta.alloc_id}` would become `${data.nomad.alloc_id}`.

Reply:
For convenience we would need to add both keys. The first one would be enough to have `${data.nomad.alloc_id}`, the second one would be used to enrich events. Fields under `"meta"` are the ones to all events.
If we only add the first one no enrichment would be made to events, and if we only add the second one we would still need to use `meta` namespace in templates (what would become `${data.meta.nomad.alloc_id}`.